In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import matplotlib.pyplot as plt #for visualizations
import seaborn as sns #for interactive visualizations
%matplotlib inline

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list the files in the input directory

import os
print(os.listdir("../input"))

# Any results you write to the current directory are saved as output.
import warnings
warnings.filterwarnings("ignore")

['train.csv', 'gender_submission.csv', 'test.csv']


In [2]:
#importing datasets
train = pd.read_csv("../input/train.csv")
test  = pd.read_csv("../input/test.csv")
totaldata = [train, test]

In [3]:
#check both train and test datasets
train.head()


PassengerId  Survived  Pclass    ...        Fare Cabin  Embarked
0            1         0       3    ...      7.2500   NaN         S
1            2         1       1    ...     71.2833   C85         C
2            3         1       3    ...      7.9250   NaN         S
3            4         1       1    ...     53.1000  C123         S
4            5         0       3    ...      8.0500   NaN         S

[5 rows x 12 columns]

In [4]:
test.head()

PassengerId  Pclass   ...    Cabin Embarked
0          892       3   ...      NaN        Q
1          893       3   ...      NaN        S
2          894       2   ...      NaN        Q
3          895       3   ...      NaN        S
4          896       3   ...      NaN        S

[5 rows x 11 columns]

In [5]:
train.tail()

PassengerId  Survived  Pclass    ...      Fare Cabin  Embarked
886          887         0       2    ...     13.00   NaN         S
887          888         1       1    ...     30.00   B42         S
888          889         0       3    ...     23.45   NaN         S
889          890         1       1    ...     30.00  C148         C
890          891         0       3    ...      7.75   NaN         Q

[5 rows x 12 columns]

In [6]:
test.tail()

PassengerId  Pclass   ...    Cabin Embarked
413         1305       3   ...      NaN        S
414         1306       1   ...     C105        C
415         1307       3   ...      NaN        S
416         1308       3   ...      NaN        S
417         1309       3   ...      NaN        C

[5 rows x 11 columns]

In [7]:
#finding columns in train data set
train.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [8]:
#the above code gives you in Index form
#this gives you in an array form
train.columns.values

array(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'], dtype=object)

In [9]:
train.columns[0]

'PassengerId'

In [10]:
train.columns.values[0]

'PassengerId'

In [11]:
#finding the shape(no.of.rows, no.of.columns) in train
train.shape

(891, 12)

In [12]:
#finding the shape(no.of.rows, no.of. columns) in test
test.shape

(418, 11)

In [13]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.6+ KB


Cabin and age has missing values, out of 12 coulmns 5 are object(text), 5 are int64(numbers) and 2 are float(in decimal points)
Name, Sex, Ticket, Cabin, Embarked are text data type.
remaining are numericals.
Age,Cabin  and Embarked has missing values.

In [14]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
PassengerId    418 non-null int64
Pclass         418 non-null int64
Name           418 non-null object
Sex            418 non-null object
Age            332 non-null float64
SibSp          418 non-null int64
Parch          418 non-null int64
Ticket         418 non-null object
Fare           417 non-null float64
Cabin          91 non-null object
Embarked       418 non-null object
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB


so we had missing values in Age,1 value in Fare and some values in Cabin

In [15]:
train.describe()

PassengerId    Survived     ...           Parch        Fare
count   891.000000  891.000000     ...      891.000000  891.000000
mean    446.000000    0.383838     ...        0.381594   32.204208
std     257.353842    0.486592     ...        0.806057   49.693429
min       1.000000    0.000000     ...        0.000000    0.000000
25%     223.500000    0.000000     ...        0.000000    7.910400
50%     446.000000    0.000000     ...        0.000000   14.454200
75%     668.500000    1.000000     ...        0.000000   31.000000
max     891.000000    1.000000     ...        6.000000  512.329200

[8 rows x 7 columns]

**Data preprocessing **

handling missing values.

from the train.info( ) we can say that we had missing values in Age, Cabin , Embarked columns as they have less rows out of 891.


In [16]:
#again checking missing values using isna() or isnull() method which most of the people do
train.isna().sum
#using this code snippet it only shows boolean values. if the column contains missing values then 
#it shows True, else it shows False.

<bound method DataFrame.sum of      PassengerId  Survived  Pclass    ...      Fare  Cabin  Embarked
0          False     False   False    ...     False   True     False
1          False     False   False    ...     False  False     False
2          False     False   False    ...     False   True     False
3          False     False   False    ...     False  False     False
4          False     False   False    ...     False   True     False
5          False     False   False    ...     False   True     False
6          False     False   False    ...     False  False     False
7          False     False   False    ...     False   True     False
8          False     False   False    ...     False   True     False
9          False     False   False    ...     False   True     False
10         False     False   False    ...     False  False     False
11         False     False   False    ...     False  False     False
12         False     False   False    ...     False   True     False
13 

In [17]:
train.isna().sum()


PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [18]:
#same code as above but using isnull() method
train.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [19]:
#from the above code snippet, its clear that we had missing values in Age, Cabin, Embarked columns.
#Cabin has 687 missing values, Age has 177 missing values, Embarked has 2 missing values.
#now try to fill those missing vlues.


In [20]:
test.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

We had 86 misssing values in Age, 327 missing values in Cabin, 1 in Fare

In [21]:
#lets replace missing values of Embarked column in trian data set
#trying to find out what and how many are the values present in Embarked column
train.Embarked.value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

In [22]:
#mostly missing values are replaced with Either of Mean,Median,Mode
#so totally we have 'S' repeated more times which is Mode case.
#so fill the missing 2 values with most repeated value i.e "S"
train["Embarked"]  = train["Embarked"].fillna("S")

In [23]:
#after filling missing values in Embarked column, verifyinig still we have missing values in 
#Embarked column or not.
train.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         0
dtype: int64

In [24]:
#so now we do not have any missing values in Embarked columns. 
#fill missing value of Age with Median.in both test and train.
train['Age'] = train['Age'].fillna(train['Age'].median())
test['Age'] = test['Age'].fillna(test['Age'].median())

In [25]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            891 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       891 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.6+ KB


In [26]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
PassengerId    418 non-null int64
Pclass         418 non-null int64
Name           418 non-null object
Sex            418 non-null object
Age            418 non-null float64
SibSp          418 non-null int64
Parch          418 non-null int64
Ticket         418 non-null object
Fare           417 non-null float64
Cabin          91 non-null object
Embarked       418 non-null object
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB


In [27]:
#lets fill the 1 missing value in Fare column of test data set
test['Fare'] = test["Fare"].fillna("0")

In [28]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
PassengerId    418 non-null int64
Pclass         418 non-null int64
Name           418 non-null object
Sex            418 non-null object
Age            418 non-null float64
SibSp          418 non-null int64
Parch          418 non-null int64
Ticket         418 non-null object
Fare           418 non-null object
Cabin          91 non-null object
Embarked       418 non-null object
dtypes: float64(1), int64(4), object(6)
memory usage: 36.0+ KB


In [29]:
#As Cabin Column has most data missing ,
#there wont be any use filling missing values as it would be hard to fill those missing values using
#using any of the techniques like Mean, Medain and  Mode.we have to delete it. 
train = train.drop("Cabin",axis = 1)
test = test.drop("Cabin", axis = 1)
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            891 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Embarked       891 non-null object
dtypes: float64(2), int64(5), object(4)
memory usage: 76.6+ KB


In [30]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 10 columns):
PassengerId    418 non-null int64
Pclass         418 non-null int64
Name           418 non-null object
Sex            418 non-null object
Age            418 non-null float64
SibSp          418 non-null int64
Parch          418 non-null int64
Ticket         418 non-null object
Fare           418 non-null object
Embarked       418 non-null object
dtypes: float64(1), int64(4), object(5)
memory usage: 32.7+ KB


In [31]:
#now we had another problem, i.e after filling all the missing values, now we have to convert text 
#data to numerical data, because that is what a machine understands at the end of the day.
#for that we have to use label encoder
#but before that lets find can we convert all the text values to numerical ones.
train.Ticket.value_counts()

1601                 7
347082               7
CA. 2343             7
3101295              6
347088               6
CA 2144              6
382652               5
S.O.C. 14879         5
2666                 4
LINE                 4
19950                4
17421                4
W./C. 6608           4
113781               4
347077               4
113760               4
349909               4
4133                 4
PC 17757             4
C.A. 31921           3
13502                3
345773               3
SC/Paris 2123        3
PC 17755             3
PC 17760             3
PC 17582             3
24160                3
110152               3
F.C.C. 13529         3
363291               3
                    ..
350060               1
12460                1
65303                1
382649               1
36967                1
C.A. 29395           1
342826               1
237789               1
A/5 3536             1
345770               1
349239               1
PC 17483             1
SC/PARIS 21

In [32]:
train['Ticket'].describe()

count      891
unique     681
top       1601
freq         7
Name: Ticket, dtype: object

In [33]:
#its clear that we had 681 unique values, so its impossible to encode it. so delete Ticket column
train = train.drop(['Ticket'],axis = 1)
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 10 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            891 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Fare           891 non-null float64
Embarked       891 non-null object
dtypes: float64(2), int64(5), object(3)
memory usage: 69.7+ KB


In [34]:
test['Ticket'].describe()

count          418
unique         363
top       PC 17608
freq             5
Name: Ticket, dtype: object

In [35]:
test = test.drop(['Ticket'],axis = 1)
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 9 columns):
PassengerId    418 non-null int64
Pclass         418 non-null int64
Name           418 non-null object
Sex            418 non-null object
Age            418 non-null float64
SibSp          418 non-null int64
Parch          418 non-null int64
Fare           418 non-null object
Embarked       418 non-null object
dtypes: float64(1), int64(4), object(4)
memory usage: 29.5+ KB


In [36]:
#also there is no use with name, so lets delete it
train = train.drop(["Name"], axis = 1)
test = test.drop(['Name'], axis =1)
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Sex            891 non-null object
Age            891 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Fare           891 non-null float64
Embarked       891 non-null object
dtypes: float64(2), int64(5), object(2)
memory usage: 62.7+ KB


In [37]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 8 columns):
PassengerId    418 non-null int64
Pclass         418 non-null int64
Sex            418 non-null object
Age            418 non-null float64
SibSp          418 non-null int64
Parch          418 non-null int64
Fare           418 non-null object
Embarked       418 non-null object
dtypes: float64(1), int64(4), object(3)
memory usage: 26.2+ KB


In [38]:
#lets encode remaning 2 text columns to numerical ones
#lets write function for label encoder
from sklearn.preprocessing import LabelEncoder

In [39]:
def encode_features(dataset,featurenames):
    for featurename in featurenames:
        LE = LabelEncoder()
        LE.fit(dataset[featurename])
        dataset[featurename] = LE.transform(dataset[featurename])
    return dataset    


In [40]:

train = encode_features(train, ['Sex','Embarked'])
test = encode_features(test, ['Sex','Embarked'])


In [41]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Sex            891 non-null int64
Age            891 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Fare           891 non-null float64
Embarked       891 non-null int64
dtypes: float64(2), int64(7)
memory usage: 62.7 KB


In [42]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 8 columns):
PassengerId    418 non-null int64
Pclass         418 non-null int64
Sex            418 non-null int64
Age            418 non-null float64
SibSp          418 non-null int64
Parch          418 non-null int64
Fare           418 non-null object
Embarked       418 non-null int64
dtypes: float64(1), int64(6), object(1)
memory usage: 26.2+ KB


In [43]:
#lets convert Fare from float to int
train['Fare'] = train['Fare'].astype(int)
test["Fare"] = test['Fare'].astype(int)
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Sex            891 non-null int64
Age            891 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Fare           891 non-null int64
Embarked       891 non-null int64
dtypes: float64(1), int64(8)
memory usage: 62.7 KB


In [44]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 8 columns):
PassengerId    418 non-null int64
Pclass         418 non-null int64
Sex            418 non-null int64
Age            418 non-null float64
SibSp          418 non-null int64
Parch          418 non-null int64
Fare           418 non-null int64
Embarked       418 non-null int64
dtypes: float64(1), int64(7)
memory usage: 26.2 KB


**Data visualization**

As our ML models fastens the processing if our data is in  0,1,2....., in simplie numeric form. so lets change Fare and Age to such format.
now we will find the correlation between Age and survival and between Fare and survival.

In [45]:
test.head()

PassengerId  Pclass  Sex   Age  SibSp  Parch  Fare  Embarked
0          892       3    1  34.5      0      0     7         1
1          893       3    0  47.0      1      0     7         2
2          894       2    1  62.0      0      0     9         1
3          895       3    1  27.0      0      0     8         2
4          896       3    0  22.0      1      1    12         2

In [46]:
train = train.drop(['PassengerId'], axis=1)
train.head()

Survived  Pclass  Sex   Age  SibSp  Parch  Fare  Embarked
0         0       3    1  22.0      1      0     7         2
1         1       1    0  38.0      1      0    71         0
2         1       3    0  26.0      0      0     7         2
3         1       1    0  35.0      1      0    53         2
4         0       3    1  35.0      0      0     8         2

**ML models**

In [47]:
X_train = train.drop(['Survived'], axis = 1)
Y_train = train["Survived"]
X_test  = test.drop("PassengerId", axis=1).copy()
X_train.shape, Y_train.shape, X_test.shape

((891, 7), (891,), (418, 7))

In [48]:
X_train.head()

Pclass  Sex   Age  SibSp  Parch  Fare  Embarked
0       3    1  22.0      1      0     7         2
1       1    0  38.0      1      0    71         0
2       3    0  26.0      0      0     7         2
3       1    0  35.0      1      0    53         2
4       3    1  35.0      0      0     8         2

In [49]:
Y_train.head()

0    0
1    1
2    1
3    1
4    0
Name: Survived, dtype: int64

In [50]:
X_test.head()

Pclass  Sex   Age  SibSp  Parch  Fare  Embarked
0       3    1  34.5      0      0     7         1
1       3    0  47.0      1      0     7         2
2       2    1  62.0      0      0     9         1
3       3    1  27.0      0      0     8         2
4       3    0  22.0      1      1    12         2

In [51]:
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier

In [52]:
#Logistic regression
LR = LogisticRegression()
LR.fit(X_train, Y_train)
y_pred_LR = LR.predict(X_test)
LR.score(X_train,Y_train)*100


80.3591470258137

In [53]:
#Support vector machine 
svm = SVC()
svm.fit(X_train, Y_train)
y_pred_svm = svm.predict(X_test)
svm.score(X_train,Y_train)*100

89.56228956228956

In [54]:
#RandomForestClassirier
rfc = RandomForestClassifier()
rfc.fit(X_train, Y_train)
y_pred_rfc = rfc.predict(X_test)
rfc.score(X_train,Y_train)*100

94.94949494949495

In [55]:
#KNeighborsClassifier
knc = KNeighborsClassifier()
knc.fit(X_train, Y_train)
y_pred_knc = knc.predict(X_test)
knc.score(X_train,Y_train)*100

81.14478114478115

In [56]:
#KNeighborsClassifier
dtc = DecisionTreeClassifier()
dtc.fit(X_train, Y_train)
y_pred_dtc = dtc.predict(X_test)
dtc.score(X_train,Y_train)*100

96.18406285072952

**If you find any mistakes or if you have any suggestions, please comment down. I would like to learn from my mistakes.**

In [57]:
submission = pd.DataFrame({
        "PassengerId": test["PassengerId"],
        "Survived": y_pred_dtc
    })
submission.to_csv('submission.csv', index=False)